### PR TITLE
app-containers: switch to upstream docker, containerd, runc, bump docker to v24

### PR DESCRIPTION
**Pull request by t-lo**
This change removes Flatcar specific builds of docker[-cli], containerd, cri-tools, and runc and instead switches to upstream Gentoo ebuilds added to portage-stable.

The change updates docker to 24.0.6, and cri-tools to 1.27.0.

**NOTE** that there currently is no upstream ebuild for containerd-1.7.7, so this change adds that ebuild based on the upstream containerd-1.7.6 ebuild.

Flatcar customisations like systemd units etc. are now applied in the manglefs script of the respective sysexts, based on file system trees in `coreos-overlay/coreos/sysext/(containerd|docker)`.

The build_sysext script has been extended by an option to strip all binaries in a sysext; the option is deactivated by default.

Lastly, the sysext command line syntax of build_image has been extended to allow specifying multiple packages for a sysext. This was necessary because docker-cli and docker do not have any runtime relationships and therefore must both be specified for installation to correctly mirror Flatcar's own docker packaging.

Closes https://github.com/flatcar/Flatcar/issues/1091 .

**Testing**
* GH CI: https://github.com/flatcar/scripts/actions/runs/6696260660/job/18193627901?pr=1305
* Jenkins CI: http://jenkins.infra.kinvolk.io:8080/job/container/job/packages_all_arches/2840/cldsv/

**Related PRs**
* https://github.com/flatcar/mantle/pull/475
* https://github.com/flatcar/mantle/pull/474

**Upstreaming**
* Upstream PR https://github.com/gentoo/gentoo/pull/33539 has been opened to add cross-compile capabilities to Gentoo's Go related eclasses.

### CODE REVIEW

File: build_library/sysext_prod_builder
Comment (t-lo) [reply build_library/sysext_prod_builder:40]: You're not annoying, quite the opposite - I really enjoy the feedback in your reviews because there's always something to learn, and the improvements are quite elegant!

File: build_library/sysext_prod_builder
Comment (t-lo) [reply build_library/sysext_prod_builder:43]: That's probably a left-over from the time where sysexts derived their names from the packages they included. Good catch!

File: build_sysext
Comment (t-lo) [reply build_sysext:237]: I don't think that exists for Go. I agree doing it here is a hacky work-around.

File: sdk_container/src/third_party/portage-stable/app-containers/containerd/containerd-1.7.7.ebuild
Comment (t-lo) [reply sdk_container/src/third_party/portage-stable/app-containers/containerd/containerd-1.7.7.ebuild:1]: Hmmm, I'll move it back to coreos-overlay then.

File: sdk_container/src/third_party/coreos-overlay/profiles/coreos/base/package.accept_keywords
Comment (t-lo) [reply sdk_container/src/third_party/coreos-overlay/profiles/coreos/base/package.accept_keywords]: Fixed.

File: changelog/updates/2023-10-25-docker-24-critools-1.27.md
Comment (t-lo) [reply changelog/updates/2023-10-25-docker-24-critools-1.27.md]: Godd point, moved.

File: build_sysext
Comment (krnowak) [reply build_sysext:237]: It's weird. I thought that portage handles it just like for any binary. But I don't have time to investigate how this stuff works exactly. Something to look into later.

File: changelog/changes/2023-10-25-docker-gentoo-upstream.md
Comment (krnowak) [inline changelog/changes/2023-10-25-docker-gentoo-upstream.md:4]: Does this affect existing installations or is it only for the new ones? Maybe it's something to add to the changelog entry?

File: sdk_container/src/third_party/portage-stable/eclass/go-env.eclass
Comment (t-lo) [reply sdk_container/src/third_party/portage-stable/eclass/go-env.eclass:26]: Good point, it's over-engineered. Will change to a simple `case` and assign `GOARCH` directly.

File: sdk_container/src/third_party/portage-stable/eclass/go-env.eclass
Comment (t-lo) [reply sdk_container/src/third_party/portage-stable/eclass/go-env.eclass:9]: Actually that can go entirely, we don't need it. Copy+paste left-over.

File: sdk_container/src/third_party/portage-stable/eclass/go-env.eclass
Comment (t-lo) [reply sdk_container/src/third_party/portage-stable/eclass/go-env.eclass:41]: Added in my suqashed commit, thanks! (I need to juggle these changes to stay in sync with https://github.com/gentoo/gentoo/pull/33539).

File: changelog/changes/2023-10-25-docker-gentoo-upstream.md
Comment (t-lo) [reply changelog/changes/2023-10-25-docker-gentoo-upstream.md:4]: It would affect existing deployments that use Docker's brtfs storage driver. I've added a more explicit note to call this out.

File: sdk_container/src/third_party/portage-stable/eclass/go-module.eclass
Comment (t-lo) [reply sdk_container/src/third_party/portage-stable/eclass/go-module.eclass]: Done - referred to the upstream PR in the comment, too.

File: sdk_container/src/third_party/portage-stable/eclass/golang-vcs-snapshot.eclass
Comment (t-lo) [reply sdk_container/src/third_party/portage-stable/eclass/golang-vcs-snapshot.eclass]: Done.

File: sdk_container/src/third_party/portage-stable/eclass/golang-vcs.eclass
Comment (t-lo) [reply sdk_container/src/third_party/portage-stable/eclass/golang-vcs.eclass]: Added.

File: build_sysext
Comment (t-lo) [reply build_sysext:237]: You were right after all, see https://github.com/moby/moby/blob/master/project/PACKAGERS.md#stripping-binaries. I'll keep the option but set it to false by default.

**Comment by t-lo** [comment]
Created https://github.com/flatcar/Flatcar/issues/1223 to track the btrfs deprecation in docker 20 -> 24 and to run tests for gauging the fallout.

Created https://github.com/flatcar/Flatcar/issues/1222 and https://github.com/flatcar/mantle/pull/474 to ensure we continue shipping the `devicemapper` docker storage driver (which _almost_ was silently discarded while working on the portage stable docker migration PR).

**Comment by t-lo** [comment]
All tests succeeded. Good to merge?

(Please also have a look at https://github.com/flatcar/mantle/pull/474 ; it's not a requirement for this PR but a nice addition)

**Comment by pothos** [comment]
Some thoughts without looking at the details:
- We should check for differences in the systemd units and the default config files (containerd toml)
- Check if the cgroupv1 containerd config still works which was created for old nodes on update
- Use this change here as opportunity to support drop-ins: https://github.com/flatcar/Flatcar/issues/1127 https://github.com/flatcar/Flatcar/issues/583

**Comment by t-lo** [comment]
> Some thoughts without looking at the details:
> 
> * We should check for differences in the systemd units and the default config files (containerd toml)

I'm actually using the systemd unit from "our" ebuild (in https://github.com/flatcar/scripts/tree/t-lo/gentoo-upstream-containerd-docker/sdk_container/src/third_party/coreos-overlay/coreos/sysext/docker/usr/lib/systemd/system); it's added to the sysext via docker's [manglefs](https://github.com/flatcar/scripts/blob/t-lo/gentoo-upstream-containerd-docker/build_library/sysext_mangle_docker-flatcar) script. 

Same for the [containerd toml](https://github.com/flatcar/scripts/tree/t-lo/gentoo-upstream-containerd-docker/sdk_container/src/third_party/coreos-overlay/coreos/sysext/containerd/usr/share/containerd) files - as far as I can tell, upstream Gentoo's containerd ebuild does not even ship any tomls.

> * Check if the cgroupv1 containerd config still works which was created for old nodes on update

Is there a test for that?

> * Use this change here as opportunity to support drop-ins: [[RFE] Define containerd drop-in config directory Flatcar#1127](https://github.com/flatcar/Flatcar/issues/1127) [[RFE] Include certs.d registry plugin in containerd config.toml Flatcar#583](https://github.com/flatcar/Flatcar/issues/583)

Absolutely, but I'd put that into a follow-up PR. Should be straightforward after this one gets merged, and could be part of the same release.

**Comment by pothos** [comment]
Ok, good to know that this here is just about the binaries :)

> Is there a test for that?

Found one kubeadm test kola, which should pass as long as the `/usr/share/` location of the cgroupv1 toml doesn't change.

**Review by pothos — COMMENTED** [review]
Could we move the containerd ebuilds to portage-stable (even if they are not really copied from Gentoo because the version was changed)? This would make it clearer how updates work in the future.

I'm also a fan of removing any ebuilds we don't use because one can accidentally use them due to wrong dependencies/use-flags.

File: sdk_container/.repo/manifests/mantle-container
Comment (pothos) [inline sdk_container/.repo/manifests/mantle-container:1]: This commit can be dropped

**Comment by t-lo** [comment]
> Could we move the containerd ebuilds to portage-stable (even if they are not really copied from Gentoo because the version was changed)? This would make it clearer how updates work in the future.
> 
> I'm also a fan of removing any ebuilds we don't use because one can accidentally use them due to wrong dependencies/use-flags.

My agreement with @krnowak is to leave these in coreos-overlay because the newest version would be deleted by the weekly updates automation. I agree we need a better way for upstreaming though.

File: sdk_container/.repo/manifests/mantle-container
Comment (t-lo) [reply sdk_container/.repo/manifests/mantle-container:1]: If I remove it then some tests will fail in my branch :) I'll rebase to main before merging anyway so this will be gone before we merge.

**Comment by t-lo** [comment]
@krnowak  @pothos re: btrfs storage backwards compatibility, this is actually built into docker. This new test succeeds: https://github.com/flatcar/mantle/pull/475

IOW if users use the btrfs storage backend and have that storage populated with at least one container, then the upgraded docker will also use btrfs storage. So I'm getting out easy on this one :)

Please have a look at https://github.com/flatcar/mantle/pull/475 and https://github.com/flatcar/mantle/pull/474, would be nice to have them merged for this PR.

**Comment by t-lo** [comment]
Rebased on latest main, restarted GH CI and Jenkins tests (and updated the test links in the summary).

File: changelog/changes/2023-10-25-docker-gentoo-upstream.md
Comment (t-lo) [reply changelog/changes/2023-10-25-docker-gentoo-upstream.md:4]: Good news: Docker includes a backwards compatibility mode for btrfs storage: If the storage is actively used, docker will automatically select the `btrfs` storage driver. I've added a [Mantle test](https://github.com/flatcar/mantle/pull/475) to validate that.

**Review by krnowak — APPROVED** [review]
So the only thing left are mantle tests, right?

**Comment by t-lo** [comment]
> So the only thing left are mantle tests, right?

Yes - the most important one is the [docker btrfs update test](https://github.com/flatcar/mantle/pull/475), the other one (devicemapper) is merged. Will wait for the open PR to be reviewed and merged, then update the mantle-container version in this PR before merging it.

**Comment by t-lo** [comment]
Updated the mantle ref to flatcar-master head; merging.

**Pull request merged**
